Wagtail Search Functionality Iteration

dlopezvsr opened the pull request:
# Description

This PR introduces the search functionality for Wagtail live pages. For this iteration we will see a simple UI, but also an autocomplete feature inside the input field powered by small jQuery/Ajax script, and simple CSS. Right now, I included the CSS and JS inside of `templates/search/search.html` template, given that on coming iterations we will modify the UI and will separate static files inside of the compiled directory.

Files added/updated:
- foundation_cms/search/views.py
- foundation_cms/search/tests/test_views.py (copied from prototype branch)
- foundation_cms/templates/search/search.html
- foundation_cms/urls.py

Link to sample test page: https://foundation-s-tp1-1628-s-lmg2pj.herokuapp.com/en/search/
Related PRs/issues: https://mozilla-hub.atlassian.net/browse/TP1-1628

**Tests**
- Go to review app [link](https://foundation-s-tp1-1628-s-lmg2pj.herokuapp.com/en/search/)
- You should see the search page and the input field.
- Start typing, and after the first 2 characters, suggestions should display.
- Confirm the locale is working correctly, and if the search is under `en` locale, only English results are showed. 

┆Issue is synchronized with this [Jira Story](https://mozilla-hub.atlassian.net/browse/TP1-2255)

danielfmiranda reviewed (approved):
Looks like everything is working as expected to me :D great job @dlopezvsr, approved!

ramram-mf commented on foundation_cms/templates/search/search.html:
I think we can achieve this result without js dependencies, what do you think @mmmavis ?

ramram-mf commented on foundation_cms/templates/search/search.html:
```diff
@@ -0,0 +1,96 @@
+{% block body_class %}template-searchresults{% endblock %}
+{% block title %}Search{% endblock %}
+{% block extra_css %}
+    <link nonce="{{ request.csp_nonce }}"
+          rel="stylesheet"
+          href="https://code.jquery.com/ui/1.12.1/themes/base/jquery-ui.css">
+    <style>
+        .ui-autocomplete {
+            max-height: 200px;
+            overflow-y: auto;
+            overflow-x: hidden;
+            padding: 5px 0;
+            background: white;
+            border: 1px solid #ddd;
+        }
+        .ui-menu-item {
+            padding: 5px 10px;
+            cursor: pointer;
+        }
+        .ui-menu-item:hover {
+            background: #f5f5f5;
+        }
+    </style>
```
```suggestion
    <style>
        .autocomplete-suggestions {
            max-height: 200px;
            overflow-y: auto;
            overflow-x: hidden;
            background: white;
            border: 1px solid #ddd;
            position: absolute;
            z-index: 1000;
            width: 30em;
        }
        .autocomplete-suggestions:empty {
            display: none;
        }
        .autocomplete-suggestion {
            padding: 5px 10px;
            cursor: pointer;
        }
        .autocomplete-suggestion:hover, .autocomplete-suggestion.active {
            background: #ccc;
        }
    </style>
```

ramram-mf commented on foundation_cms/templates/search/search.html:
```diff
@@ -0,0 +1,96 @@
+               placeholder="Search..."
+               autocomplete="off">
+        <input type="submit" value="Search" class="button">
+    </form>
```
```suggestion
        <div id="autocomplete-container" class="autocomplete-suggestions"></div>
    </form>
```

ramram-mf commented on foundation_cms/templates/search/search.html:
```diff
@@ -0,0 +1,96 @@
+    {% endif %}
+{% endblock %}
+{% block extra_js %}
+    <script nonce="{{ request.csp_nonce }}"
+            src="https://code.jquery.com/jquery-3.6.0.min.js"></script>
+    <script nonce="{{ request.csp_nonce }}"
+            src="https://code.jquery.com/ui/1.12.1/jquery-ui.min.js"></script>
+    <script nonce="{{ request.csp_nonce }}">
+        $(document).ready(function() {
+            $("#search-input").autocomplete({
+                source: function(request, response) {
+                    $.ajax({
+                        url: "{% url 'search_autocomplete' %}",
+                        dataType: "json",
+                        data: {
+                            query: request.term
+                        },
+                        success: function(data) {
+                            response($.map(data.results, function(item) {
+                                return {
+                                    label: item.title,
+                                    value: item.title,
+                                    url: item.url
+                                };
+                            }));
+                        }
+                    });
+                },
+                minLength: 2,
+                select: function(event, ui) {
+                    window.location.href = ui.item.url;
+                    return false;
+                }
+            });
+        });
+    </script>
```
```suggestion
    <script nonce="{{ request.csp_nonce }}">
        document.addEventListener('DOMContentLoaded', function() {
            const searchInput = document.getElementById('search-input');
            const autocompleteContainer = document.getElementById('autocomplete-container');
            let activeIndex = -1;

            searchInput.addEventListener('input', async function() {
                const query = searchInput.value.trim();
                if (query.length < 2) {
                    autocompleteContainer.innerHTML = '';
                    activeIndex = -1;
                    return;
                }

                try {
                    const response = await fetch("{% url 'search_autocomplete' %}?query=" + encodeURIComponent(query));
                    if (!response.ok) {
                        console.error('Error fetching autocomplete results');
                        return;
                    }

                    const data = await response.json();
                    autocompleteContainer.innerHTML = '';
                    activeIndex = -1;

                    data.results.forEach(item => {
                        const suggestion = document.createElement('div');
                        suggestion.className = 'autocomplete-suggestion';
                        suggestion.textContent = item.title;
                        suggestion.addEventListener('click', () => {
                            window.location.href = item.url;
                        });
                        autocompleteContainer.appendChild(suggestion);
                    });
                } catch (error) {
                    console.error('Error:', error);
                }
            });

            searchInput.addEventListener('keydown', function(event) {
                const suggestions = autocompleteContainer.querySelectorAll('.autocomplete-suggestion');
                if (suggestions.length === 0) return;

                if (event.key === 'ArrowDown') {
                     event.preventDefault();
                    activeIndex = (activeIndex + 1) % suggestions.length;
                    updateActiveSuggestion(suggestions);
                } else if (event.key === 'ArrowUp') {
                    event.preventDefault();
                    activeIndex = (activeIndex - 1 + suggestions.length) % suggestions.length;
                    updateActiveSuggestion(suggestions);
                } else if (event.key === 'Enter') {
                    event.preventDefault();
                    if (activeIndex >= 0 && activeIndex < suggestions.length) {
                        const activeSuggestion = suggestions[activeIndex];
                        window.location.href = activeSuggestion.dataset.url;
                    }
                } else if (event.key === 'Escape') {
                    autocompleteContainer.innerHTML = '';
                    activeIndex = -1;
                }
            });

            document.addEventListener('click', function(event) {
                if (!autocompleteContainer.contains(event.target) && event.target !== searchInput) {
                    autocompleteContainer.innerHTML = '';
                    activeIndex = -1;
                }
            });

            function updateActiveSuggestion(suggestions) {
                suggestions.forEach((suggestion, index) => {
                    if (index === activeIndex) {
                        suggestion.classList.add('active');
                        suggestion.scrollIntoView({ block: 'nearest' });
                    } else {
                        suggestion.classList.remove('active');
                    }
                });
            }
        });
    </script>
```

ramram-mf reviewed (changes requested):
I've added some suggestions that seem to replicate the features jquery and ui brought to the table. I think we can start from here dependency free.

dlopezvsr commented:
> I've added some suggestions that seem to replicate the features jquery and ui brought to the table. I think we can start from here dependency free.

Thanks so much @ramram-mf I just commited your suggestions, they worked perfectly. I also removed values from `CSP_SCRIPT_SRC` in Heroku.
